File: Data-Structures/week4/assignment/phone_book/phone_book.py
# python3

class Query:
	def __init__(self, query):
		self.type = query[0]
		self.number = int(query[1])
		if self.type == 'add':
			self.name = query[2]

def read_queries():
	n = int(input())
	return [Query(input().split()) for i in range(n)]

def write_responses(result):
	print('\n'.join(result))

def process_queries(queries):
	result = []
	# Keep list of all existing (i.e. not deleted yet) contacts.
	contacts = {}
	for cur_query in queries:
		if cur_query.type == 'add':
			# if we already have contact with such number,
			# we should rewrite contact's name
			contacts[cur_query.number] = cur_query.name
		elif cur_query.type == 'del':
			if cur_query.number in contacts:
				del contacts[cur_query.number]
		else:
			response = 'not found' if cur_query.number not in contacts else contacts[cur_query.number]
			result.append(response)
	return result

if __name__ == '__main__':
	write_responses(process_queries(read_queries()))

File: coursera-Data-Structures-and-Algorithms-specialization/Data-Structures/week4/assignment/phone_book/PhoneBook.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.StringTokenizer;
import java.util.HashMap;

public class PhoneBook {

	private FastScanner in = new FastScanner();
	// Keep list of all existing (i.e. not deleted yet) contacts.
	private HashMap<Integer, String> contacts = new HashMap<Integer, String>();

	public static void main(String[] args) {
		new PhoneBook().processQueries();
	}

	private Query readQuery() {
		String type = in.next();
		int number = in.nextInt();
		if (type.equals("add")) {
			String name = in.next();
			return new Query(type, name, number);
		} else {
			return new Query(type, number);
		}
	}

	private void writeResponse(String response) {
		System.out.println(response);
	}


	private void processQuery(Query query) {
		if (query.type.equals("add")) {
			// if we already have contact with such number,
			// we should rewrite contact's name
			contacts.put(query.number, query.name);
		} else if (query.type.equals("del")) {
			if (contacts.containsKey(query.number)) contacts.remove(query.number);
		} else {
			String response = "not found";
			if (contacts.containsKey(query.number)) response = contacts.get(query.number);
			writeResponse(response);
		}
	}

	public void processQueries() {
		int queryCount = in.nextInt();
		for (int i = 0; i < queryCount; ++i)
			processQuery(readQuery());
	}

	static class Contact {
		String name;
		int number;

		public Contact(String name, int number) {
			this.name = name;
			this.number = number;
		}
	}

	static class Query {
		String type;
		String name;
		int number;

		public Query(String type, String name, int number) {
			this.type = type;
			this.name = name;
			this.number = number;
		}

		public Query(String type, int number) {
			this.type = type;
			this.number = number;
		}
	}

	class FastScanner {
		BufferedReader br;
		StringTokenizer st;

		FastScanner() {
			br = new BufferedReader(new InputStreamReader(System.in));
		}

		String next() {
			while (st == null || !st.hasMoreTokens()) {
				try {
					st = new StringTokenizer(br.readLine());
				} catch (IOException e) {
					e.printStackTrace();
				}
			}
			return st.nextToken();
		}

		int nextInt() {
			return Integer.parseInt(next());
		}
	}
}
